packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m

```diff
@@ -0,0 +1,440 @@
+                    reject:(UMPromiseRejectBlock)reject)
+{
+  BOOL promiseSet = [self setPromise:QUERY_HISTORY_KEY resolve:resolve reject:reject];
+
+  if(promiseSet) {
+    // Request history
+    [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
+  }
+}
+
+UM_EXPORT_METHOD_AS(disconnectAsync,
+                    disconnectAsync:(UMPromiseResolveBlock)resolve
+                    reject:(UMPromiseRejectBlock)reject)
+{
+  [[SKPaymentQueue defaultQueue] removeTransactionObserver:self];
+  resolve(nil);
+}
+
+# pragma mark - Helper Methods
+
+- (void)requestProducts:(NSArray *)productIdentifiers
+{
+  SKProductsRequest *productsRequest = [[SKProductsRequest alloc]
+                                        initWithProductIdentifiers:[NSSet setWithArray:productIdentifiers]];
+  // Keep a strong reference to the request
+  _request = productsRequest;
+
+  // Here we are the delegate since this class also implements SKProductsRequestDelegate
+  productsRequest.delegate = self;
+
+  // This will return in the productsRequest method below
+  [productsRequest start];
+}
+
+- (void)handleQuery:(SKProductsResponse *)response {
+  NSMutableArray *result = [NSMutableArray array];
+
+  for (SKProduct *validProduct in response.products) {
+    NSDictionary *productData = [self getProductData:validProduct];
+    [result addObject:productData];
+  }
+
+  _queryingItems = NO;
+  NSDictionary *res = [self formatResults:result withResponseCode:OK];
+  [self resolvePromise:QUERY_PURCHASABLE_KEY value:res];
+}
+
+-(void)handlePurchase:(SKProductsResponse *)response {
+  for (NSString *invalidIdentifier in response.invalidProductIdentifiers) {
+    NSDictionary *results = [self formatResults:SKErrorStoreProductNotAvailable];
+    [self resolvePromise:invalidIdentifier value:results];
+  }
+
+  for (SKProduct *validProduct in response.products) {
+    [self purchase:validProduct];
+  }
+}
+
+- (void)purchase:(SKProduct *)product
+{
+  SKPayment *payment = [SKPayment paymentWithProduct:product];
+  [[SKPaymentQueue defaultQueue] addPayment:payment];
+}
+
+# pragma mark - StoreKit Transaction Observer Methods
+
+/*
+ This function is called both when purchasing an item and querying for item data
+ */
+- (void)productsRequest:(SKProductsRequest *)request didReceiveResponse:(SKProductsResponse *)response
+{
+  if (_queryingItems) {
+    [self handleQuery:response];
+  } else {
+    [self handlePurchase:response];
+  }
+}
+
+- (void)paymentQueueRestoreCompletedTransactionsFinished:(SKPaymentQueue *)queue
+{
+  NSMutableArray *results = [NSMutableArray array];
+
+  for (SKPaymentTransaction *transaction in queue.transactions) {
+    SKPaymentTransactionState transactionState = transaction.transactionState;
+    if (transactionState == SKPaymentTransactionStateRestored || transactionState == SKPaymentTransactionStatePurchased) {
+      NSDictionary * transactionData = [self getTransactionData:transaction];
```

Comment:
`getTransactionData` hardcodes `acknowledged` to be `NO`, while this method calls it and then finishes the transaction (line 210). This means that the JS side observes `acknowledged: false`, but the transaction has been marked finished by that time.

Reply:
Yeah my reasoning for always returning `NO` is because I want the user to always finish the transaction when it's returned. This shouldn't be an issue in the purchase listener since every time the purchase is emitted from the listener it means it came from the transaction queue, therefore `finishTransaction` has not been called yet. 

If you're talking about checking the history returned from `connectAsync` as we discussed in the above example that should be fine as well since calling `finishTransaction` on an already finished transaction is a no-op and doesn't cause any issues. See:
```Objective-C
// transaction would be nil in this case
SKPaymentTransaction *transaction = _pendingTransactions[purchaseToken];
_pendingTransactions[purchaseToken] = nil;
if (transaction != nil) {
  [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
}
resolve(nil);
```

The problem is there is no easy way to determine whether a transaction is "acknowledged" since iOS doesn't have this concept in the same way Android does. I could store the acknowledged purchases as a set and check against that, but that would be lost when you exit/kill the app and reopen it which is when you'd want to call `connectAsync` anyways.

Reply:
Anyways, fixed in https://github.com/expo/expo/pull/4416/commits/b9679fe3922961a252f346cc1e60a0f246595f28.

Now all of the transactions returned in `connectAsync` and `getPurchaseHistoryAsync` have `acknowledged: true` meanwhile those emitted to `onPurchase` by the event emitter have `acknowledged: false` as it should be.